File: 2023/ocaml/day03b/day03b.ml
open Printf

type position = { row : int; col : int; }

module PositionMap = Map.Make(struct
  type t = position
  let compare p1 p2 =
    match compare p1.row p2.row with
    | 0 -> compare p1.col p2.col
    | result -> result
end)

let usage () =
  let progname = Sys.argv.(0) in
  printf "usage: %s <file>\n" progname;
  exit 1

let is_digit ch =
  match ch with
    '0'..'9' -> true
  | _ -> false

let build_numbers contents =
  let (_, number_locs) = List.fold_left (fun (row, number_locs) line ->
    let (_, _, _, _, num_locs) = List.fold_left (fun (col, scanning_number, number, current_pos, num_locs) ch ->
      if scanning_number
        then
          if is_digit ch
            then
              if (col + 1) = (String.length line)
                then (col + 1, false, "", { row = -1; col = -1 }, (PositionMap.add current_pos (number ^ (String.make 1 ch)) num_locs))
                else (col + 1, true, (number ^ (String.make 1 ch)), current_pos, num_locs)
            else (col + 1, false, "", { row = -1; col = -1 }, (PositionMap.add current_pos number num_locs))
        else
          if is_digit ch
            then (col + 1, true, (String.make 1 ch), { row = row; col = col }, num_locs)
            else (col + 1, false, number, current_pos, num_locs)
      ) (0, false, "", { row = -1; col = -1 }, number_locs) (List.init (String.length line) (String.get line))
    in (row + 1, num_locs)
    ) (0, PositionMap.empty) (String.split_on_char '\n' contents)
  in number_locs

let build_parts contents =
  let (_, part_locs) = List.fold_left (fun (row, part_locs) line ->
    let (_, prt_locs) = List.fold_left (fun (col, prt_locs) ch ->
      if ch = '*'
        then (col + 1, (PositionMap.add { row = row; col = col } ch prt_locs))
        else (col + 1, prt_locs)
      ) (0, part_locs) (List.init (String.length line) (String.get line))
    in (row + 1, prt_locs)
    ) (0, PositionMap.empty) (String.split_on_char '\n' contents)
  in part_locs

let check_parts number_locs part_locs =
  List.fold_left (fun result part_loc ->
    let adjacents = List.fold_left (fun adjacents_list (number_loc, number) ->
        let number_row = number_loc.row in
        let number_col_first = number_loc.col in
        let number_col_last = number_loc.col + (String.length number) - 1 in
        let neighbors = [
          { row = -1; col = -1 }; { row = -1; col = 0 }; { row = -1; col = 1 };
          { row = 0; col = -1 }; { row = 0; col = 1 };
          { row = 1; col = -1 }; { row = 1; col = 0 }; { row = 1; col = 1 }    
        ] in
        let found_adjacent = List.fold_left (fun found neighbor ->
          let adjacent_pos = { row = part_loc.row + neighbor.row; col = part_loc.col + neighbor.col }
          in found || ((adjacent_pos.row == number_row) &&
                       (adjacent_pos.col >= number_col_first) &&
                       (adjacent_pos.col <= number_col_last))
        ) false neighbors
        in
          if found_adjacent
            then adjacents_list @ [int_of_string number]
            else adjacents_list
      ) [] (PositionMap.bindings number_locs)
    in
      if (List.length adjacents) = 2
        then result + (List.fold_left (fun acc num -> acc * num) 1 adjacents)
        else result
  ) 0 (List.map (fun (k, v) -> k) (PositionMap.bindings part_locs))

let process contents =
  let number_locs = build_numbers contents in
  let part_locs = build_parts contents in
  check_parts number_locs part_locs

let () =
  if Array.length Sys.argv < 2 then
    usage ()
  else
    let filename = Sys.argv.(1) in
    let input_channel = open_in filename in
    try
      let contents = In_channel.input_all input_channel in
      let result = process contents in
        printf "result = %d\n" result
    with e ->
      raise e

File: 2023/ruby/day03b/day03b.rb
#!/usr/bin/env ruby

Position = Struct.new(:row, :col)

def usage
  progname = $PROGRAM_NAME
  puts "usage: #{progname} <file>"
  exit(1)
end

def build_numbers(contents)
  number_locs = {}
  scanning_number = false
  number = ""
  current_pos = Position.new(-1, -1)
  contents.each_line.with_index do |line, row|
    line.strip().each_char.with_index do |ch, col|
      if scanning_number
        if ch.match?(/\d/)
          number = number + ch
        else
          number_locs[current_pos] = number
          number = ""
          scanning_number = false
        end
      else
        if ch.match?(/\d/)
          number = number + ch
          current_pos = Position.new(row, col)
          scanning_number = true
        end
      end
    end
    if scanning_number
      number_locs[current_pos] = number
      number = ""
      scanning_number = false
    end
  end
  number_locs
end

def build_parts(contents)
  part_locs = {}
  contents.each_line.with_index do |line, row|
    line.strip().each_char.with_index do |ch, col|
      if !ch.match?(/\d/) && ch != '.'
        part_locs[Position.new(row, col)] = ch
      end
    end
  end
  part_locs
end

def check_parts(number_locs, part_locs)
  result = 0
  neighbors = [
    Position.new(-1, -1), Position.new(-1, 0), Position.new(-1, 1),
    Position.new(0, -1), Position.new(0, 1),
    Position.new(1, -1), Position.new(1, 0), Position.new(1, 1)
  ]
  part_locs.each_key do |part_loc|
    adj_count = 0
    prod = 1
    number_locs.each do |num_loc, number|
      num_row = num_loc.row
      num_col_first = num_loc.col
      num_col_last = num_loc.col + number.length - 1
      found = false
      neighbors.each do |neighbor|
        adjacent_pos = Position.new(part_loc.row + neighbor.row, part_loc.col + neighbor.col)
        (num_col_first..num_col_last).each do |num_col|
          if (num_row == adjacent_pos.row) && (num_col == adjacent_pos.col)
            adj_count += 1
            prod *= number.to_i
            found = true
            break
          end
        end
        if found
          break
        end
      end
    end
    if adj_count == 2
      result += prod
    end
  end
  result
end

def process(contents)
  number_locs = build_numbers(contents)
  part_locs = build_parts(contents)
  check_parts(number_locs, part_locs)
end
  
def main()
  if ARGV.length < 1
    usage
  end
  
  filename = ARGV[0]
  contents = File.read(filename)
  result = process(contents)
  puts "result = #{result}"
end

main()
